File: src/main/java/io/github/aalsanie/codes/MappingResult.java
package io.github.aalsanie.codes;

import org.jspecify.annotations.Nullable;

import java.util.Objects;
import java.util.function.Function;
import java.util.function.Supplier;

public abstract sealed class MappingResult<T> permits MappingResult.Mapped, MappingResult.Unmapped {
    private MappingResult() {
    }

    public static <T> MappingResult<T> mapped(T value) {
        return new Mapped<>(Objects.requireNonNull(value, "value"));
    }

    @SuppressWarnings("unchecked")
    public static <T> MappingResult<T> unmapped() {
        return (MappingResult<T>) Unmapped.INSTANCE;
    }

    public abstract boolean isMapped();

    public final boolean isUnmapped() {
        return !isMapped();
    }

    public abstract @Nullable T orNull();

    public final <R> R fold(Function<? super T, ? extends R> onMapped, Supplier<? extends R> onUnmapped) {
        Objects.requireNonNull(onMapped, "onMapped");
        Objects.requireNonNull(onUnmapped, "onUnmapped");
        if (this instanceof Mapped<T> mapped) {
            return onMapped.apply(mapped.value);
        }
        return onUnmapped.get();
    }

    public static final class Mapped<T> extends MappingResult<T> {
        private final T value;

        private Mapped(T value) {
            this.value = value;
        }

        public T getValue() {
            return value;
        }

        @Override
        public boolean isMapped() {
            return true;
        }

        @Override
        public @Nullable T orNull() {
            return value;
        }

        @Override
        public boolean equals(Object other) {
            return this == other || (other instanceof Mapped<?> mapped && value.equals(mapped.value));
        }

        @Override
        public int hashCode() {
            return value.hashCode();
        }

        @Override
        public String toString() {
            return "Mapped(value=" + value + ")";
        }
    }

    public static final class Unmapped<T> extends MappingResult<T> {
        private static final Unmapped<?> INSTANCE = new Unmapped<>();

        private Unmapped() {
        }

        @Override
        public boolean isMapped() {
            return false;
        }

        @Override
        public @Nullable T orNull() {
            return null;
        }

        @Override
        public boolean equals(Object other) {
            return this == other;
        }

        @Override
        public int hashCode() {
            return 0;
        }

        @Override
        public String toString() {
            return "Unmapped";
        }
    }
}

File: src/main/kotlin/io/github/aalsanie/codes/MappingResult.kt
package io.github.aalsanie.codes

sealed class MappingResult<out T : Any> private constructor() {
    class Mapped<T : Any> internal constructor(val value: T) : MappingResult<T>() {
        override fun equals(other: Any?): Boolean =
            this === other || other is Mapped<*> && value == other.value

        override fun hashCode(): Int = value.hashCode()

        override fun toString(): String = "Mapped(value=$value)"
    }

    data object Unmapped : MappingResult<Nothing>()

    val isMapped: Boolean
        get() = this is Mapped<*>

    val isUnmapped: Boolean
        get() = this === Unmapped

    fun orNull(): T? =
        when (this) {
            is Mapped -> value
            Unmapped -> null
        }

    inline fun <R> fold(onMapped: (T) -> R, onUnmapped: () -> R): R =
        when (this) {
            is Mapped -> onMapped(value)
            Unmapped -> onUnmapped()
        }

    companion object {
        @JvmStatic
        fun <T : Any> mapped(value: T): MappingResult<T> = Mapped(value)

        @JvmStatic
        fun <T : Any> unmapped(): MappingResult<T> = Unmapped
    }
}
